core/src/avm2/stack.rs

```diff
@@ -0,0 +1,186 @@
+//! Internal representation of the AVM2 value stack
+
+use crate::avm2::function::FunctionArgs;
+use crate::avm2::method::Method;
+use crate::avm2::value::Value;
+
+use gc_arena::collect::Trace;
+use gc_arena::{Collect, Gc, Mutation};
+use std::cell::Cell;
+
+const PREALLOCATED_STACK_SIZE: usize = 200000;
+
+/// The global, preallocated value stack. This has little use directly due to
+/// the way AVM2 works; to do anything with it, an StackFrame must be
+/// obtained first through the `get_stack_slice` method.
+///
+/// We use this instead of a `Vec<Value>` to allow for obtaining "mutable" (via
+/// interior mutability) references to the stack without requiring a mutable
+/// borrow on the stack or the `Avm2`.
+#[derive(Clone, Collect, Copy)]
+#[collect(no_drop)]
+pub struct Stack<'gc>(Gc<'gc, StackData<'gc>>);
+
+struct StackData<'gc> {
+    /// Stack data
+    stack: Box<[Cell<Value<'gc>>]>,
+
+    /// The number of values currently stored on the stack
+    stack_pointer: Cell<usize>,
+}
+
+impl<'gc> Stack<'gc> {
+    pub fn new(mc: &Mutation<'gc>) -> Self {
+        let stack = (0..PREALLOCATED_STACK_SIZE)
+            .map(|_| Cell::new(Value::Undefined))
+            .collect::<Box<[_]>>();
+
+        Stack(Gc::new(
+            mc,
+            StackData {
+                stack,
+                stack_pointer: Cell::new(0),
+            },
+        ))
+    }
+
+    /// Returns a slice of stack data for the specified method, starting at the
+    /// current stack pointer. Stack frames obtained from this method must be
+    /// properly disposed of by using the `dispose_stack_frame` method.
+    pub fn get_stack_frame(&self, method: Method<'gc>) -> StackFrame<'_, 'gc> {
+        // First calculate the frame size
+        let body = method
+            .body()
+            .expect("Cannot execute non-native method without body");
+        let frame_size = body.max_stack as usize + body.num_locals as usize;
+
+        // Then actually create the stack frame
+        let stack_data = &self.0.stack;
+        let stack_pointer = &self.0.stack_pointer;
+
+        let subslice = &stack_data[stack_pointer.get()..stack_pointer.get() + frame_size];
+
+        stack_pointer.set(stack_pointer.get() + frame_size);
+
+        // Ensure the StackFrame returned to the caller does not have any
+        // old values on it, as these may contain Gc pointers that have already
+        // been collected.
+        for value in subslice {
+            value.set(Value::Undefined);
+        }
+
+        StackFrame::for_data(subslice)
+    }
+
+    pub fn dispose_stack_frame(&self, stack_frame: StackFrame<'_, 'gc>) {
```

Comment:
IMO a bit pointless to take it by move if the only place we call this uses `.copied()`. How about either passing a `&`, or if you want it to be super safe then maybe change the `.copied()` into `.take()` that sets the values to empty to prevent accidental reuse?
(but that is additional work each call for a tiny safety measure so up to you)